tests/func-tests/node_placement_test.go

```diff
@@ -7,49 +7,129 @@ import (
-	"kubevirt.io/client-go/kubecli"
-	"kubevirt.io/kubevirt/tests/flags"
-
 	tests "github.com/kubevirt/hyperconverged-cluster-operator/tests/func-tests"
+	"kubevirt.io/client-go/kubecli"
 )
 
-const nodePlacementSamplingSize = 20
-
-var _ = Describe("[rfe_id:4356][crit:medium][vendor:cnv-qe@redhat.com][level:system]Node Placement", func() {
+const (
+	hcoLabel    = "node.kubernetes.io/hco-test-node-type"
+	infra       = "infra"
+	workloads   = "workloads"
+	hcoResource = "hyperconvergeds"
+)
 
+var _ = Describe("[rfe_id:4356][crit:medium][vendor:cnv-qe@redhat.com][level:system]Node Placement", Ordered, func() {
 	var workloadsNode *v1.Node
-
+	ctx := context.TODO()
 	tests.FlagParse()
 	client, err := kubecli.GetKubevirtClient()
 	kvtutil.PanicOnError(err)
 
-	workloadsNodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{
-		LabelSelector: "node.kubernetes.io/hco-test-node-type==workloads",
-	})
-	kvtutil.PanicOnError(err)
+	BeforeAll(func() {
+		nodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{LabelSelector: "node-role.kubernetes.io/control-plane!="})
+		kvtutil.PanicOnError(err)
+		if len(nodes.Items) < 2 {
+			Skip("Skipping Node Placement tests due to insufficient cluster nodes")
+		}
+		// Label all but first node with "node.kubernetes.io/hco-test-node-type=infra"
+		// We are doing this to remove dependency of this Describe block on a shell script that
+		// labels the nodes this way
+		for _, node := range nodes.Items[:len(nodes.Items)-1] {
+			err = setHcoNodeTypeLabel(client, &node, infra)
+			kvtutil.PanicOnError(err)
+		}
+		// Label the last node with "node.kubernetes.io/hco-test-node-type=workloads"
+		err = setHcoNodeTypeLabel(client, &nodes.Items[len(nodes.Items)-1], workloads)
+		kvtutil.PanicOnError(err)
+
+		// modify the HCO CR to use the labels we just applied to the nodes
+		originalHco := tests.GetHCO(ctx, client)
+		// modify the "infra" and "workloads" keys
+		infraVal := hcov1beta1.HyperConvergedConfig{
+			NodePlacement: &sdkapi.NodePlacement{
+				NodeSelector: map[string]string{hcoLabel: infra},
+			},
+		}
+		workloadsVal := hcov1beta1.HyperConvergedConfig{
+			NodePlacement: &sdkapi.NodePlacement{
+				NodeSelector: map[string]string{hcoLabel: workloads},
+			},
+		}
+
+		hco := &hcov1beta1.HyperConverged{}
+		originalHco.DeepCopyInto(hco)
+
+		hco.Spec.Infra = infraVal
+		hco.Spec.Workloads = workloadsVal
+		hco.SetGroupVersionKind(schema.GroupVersionKind{
+			Group:   util.APIVersionGroup,
+			Version: util.CurrentAPIVersion,
+			Kind:    util.HyperConvergedKind,
+		})
+		tests.UpdateHCORetry(ctx, client, hco)
 
-	if workloadsNodes != nil && len(workloadsNodes.Items) == 1 {
+		workloadsNodes, err := client.CoreV1().Nodes().List(context.TODO(), k8smetav1.ListOptions{
+			LabelSelector: "node.kubernetes.io/hco-test-node-type==workloads",
+		})
+		kvtutil.PanicOnError(err)
 		workloadsNode = &workloadsNodes.Items[0]
```

Comment:
Please check `workloadsNodes` before accessing a specific index. Just for case.

Reply:
The reason I've not done that is because:
1. The following check right after getting nodes ensures we are not running on too few nodes:
    ```go
    if len(nodes.Items) < 2 {
	    Skip("Skipping Node Placement tests due to insufficient cluster nodes")
    }
    ```
2. Following code which always labels the last node with the label for workload nodes:
    ```go
    // Label the last node with "node.kubernetes.io/hco-test-node-type=workloads"
    err = setHcoNodeTypeLabel(client, &nodes.Items[len(nodes.Items)-1], workloads)
    ```

So it's kinda implicit.

If you'd rather it be explicit, I am OK to make the change. I just wanted to point that, in my understanding, we are covered because there's always only one node with that label and the code earlier in the function covers the part that we have at least two cluster nodes besides the control-plane node(s).

WDYT?

Reply:
It's more for safety. To not access index 0 before we know it's there

Reply:
Done.